image/image_linux.go

```diff
@@ -688,33 +595,248 @@ var setupSeed = func(tsto *tooling.ToolingStore, model *asserts.Model, opts *Opt
+		return err
+	}
+
+	copySnap := func(name, src, dst string) error {
+		fmt.Fprintf(Stdout, "Copying %q (%s)\n", src, name)
+		return osutil.CopyFile(src, dst, 0)
+	}
+	if err := s.w.SeedSnaps(copySnap); err != nil {
+		return err
+	}
+
+	if err := s.w.WriteMeta(); err != nil {
+		return err
+	}
+
+	// TODO: There will be classic UC20+ model based systems
+	//       that will have a bootable  ubuntu-seed partition.
+	//       This will need to be handled here eventually too.
+	if s.classic {
+		return s.finishSeedClassic()
+	}
+	return s.finishSeedCore()
+}
+
+func optionSnaps(opts *Options) []*seedwriter.OptionsSnap {
+	optSnaps := make([]*seedwriter.OptionsSnap, 0, len(opts.Snaps))
+	for _, snapName := range opts.Snaps {
+		var optSnap seedwriter.OptionsSnap
+		if strings.HasSuffix(snapName, ".snap") {
+			// local
+			optSnap.Path = snapName
+		} else {
+			optSnap.Name = snapName
+		}
+		optSnap.Channel = opts.SnapChannels[snapName]
+		optSnaps = append(optSnaps, &optSnap)
+	}
+	return optSnaps
+}
+
+func manifestFromLocalSnaps(snaps localSnapRefs, opts *Options) (map[string]snap.Revision, error) {
+	// Build a map of snaps for the manifest file
+	seedManifest := make(map[string]snap.Revision)
+
+	// Check local snaps again, but now after InfoDerived has been called. InfoDerived
+	// fills out the snap revisions for the local snaps, and we need this to verify against
+	// expected revisions.
+	for sn := range snaps {
+		// Its a bit more tricky to deal with local snaps, as we only have that specific revision
+		// available. Therefore the revision in the local snap must be exactly the revision specified
+		// in the manifest. If it's not, we fail.
+		specifiedRevision := opts.Revisions[sn.Info.SnapName()]
+		if !specifiedRevision.Unset() && specifiedRevision != sn.Info.Revision {
+			return nil, fmt.Errorf("cannot use snap %s for image, unknown/local revision does not match the value specified by revisions file (%s != %s)",
+				sn.Path, sn.Info.Revision, specifiedRevision)
+		}
+		if !sn.Info.Revision.Unset() {
+			seedManifest[sn.Info.SnapName()] = sn.Info.Revision
+		}
+	}
+	return seedManifest, nil
+}
+
+func selectAssertionMaxFormats(tsto *tooling.ToolingStore, sysSn *seedwriter.SeedSnap) error {
+	if sysSn == nil {
+		// nothing to do
+		return nil
+	}
+	snapf, err := snapfile.Open(sysSn.Path)
+	if err != nil {
+		return err
+	}
+	// XXX take also kernel into account
+	// XXX warning logic
+	maxFormats, _, err := snap.SnapdAssertionMaxFormatsFromSnapFile(snapf)
+	if err != nil {
+		return err
+	}
+	tsto.SetAssertionMaxFormats(maxFormats)
+	return nil
+}
+
+var setupSeed = func(tsto *tooling.ToolingStore, model *asserts.Model, opts *Options) error {
+	s, err := newImageSeeder(tsto, model, opts)
+	if err != nil {
+		return err
+	}
+
+	// TODO: developer database in home or use snapd (but need
+	// a bit more API there, potential issues when crossing stores/series)
+	db, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
```

Comment:
Perhaps move OpenDatabase bit into s.start() ?

Reply:
I would not do that, imageSeeder is not keeping the db directly, it also makes Writer.Start and imageSeeder.Start diverge further

Reply:
I'll leave this as-is